### DISCUSSION TITLE
Cleaning up the cache

### DISCUSSION THREAD

**Discussion by eidolonFIRE** [JaffaKetchup/flutter_map_tile_caching#62] [Q&A] [answered]
I am reading in the docs that setting the `maxStoreLength` can be performance expensive when the limit is very high because it may need to sort through many thousands of files to delete the oldest. It also appears from the description that `cachedValidDuration` doesn't affect the retention of the files, only forces a cache miss for a given tile.

**Question:**  Can I voluntarily scan through the cache to purge all files dated before a given DateTime? It would be great if I could do this occasionally rather than set a hard tile count limit.

**Comment by JaffaKetchup** [comment, marked as answer]
Hi there @eidolonFIRE.
You will have to do this manually through 'access.tiles'. It contains a file for each tile. After deleting the appropriate files, make sure you use [`invalidateCachedStatistics()`](https://fmtc.jaffaketchup.dev/usage/roots-and-stores/statistics#statistic-caching).
Let me know if you need more help!

**Comment by eidolonFIRE** [comment]
I can't get `store.stats.invalidateCachedStatistics();` to actually do anything. Tried every combination of running and waiting or even restarting the app. I peek at the files and they don't change either.

Adding `.noCache` before getting the store size gives me the value I expect.

**Reply by JaffaKetchup** [reply]
Hmm, what platform are you running on? Additionally, are you sure you are looking in the 'statistics' directory?

Try calling it with 'null'. This should clear all values. Check if cache hits and misses reset to 0 afterward.

**Reply by JaffaKetchup** [reply]
Also, note that you should append 'Async' to getter and method names wherever possible to improve performance.

**Reply by eidolonFIRE** [reply]
Android Emulator, latest commit on `main` branch.

I have confirmed I'm successfully deleting the tiles that I want to. I can double check in `stores/<name>/tiles` and verify.
For the stats, looking in `stores/<name>/stats`. I see the `size.cache` file which contains a `double` value. 

My function to selectively clean the tiles actually looks basically the same as `store.reset`

I am able to reproduce my bug reliably now... `invalidateCachedStatistics` and `reset` doesn't do anything on a given store if it isn't the one currently active in my flutter map. I have multiple stores, each pointing to a different URL (different map base layers).
My setup with leaflet is very similar to the example project.

Everything works perfectly until I've switched from one store to another... then none of them clear their stats unless they are actively used in leaflet and I scroll around in the map.

**Reply by JaffaKetchup** [reply]
Can you try the example application? It seems to work for me.

**Reply by JaffaKetchup** [reply]
@eidolonFIRE Managed to fix your issues? Happy to help more if necessary.

**Reply by eidolonFIRE** [reply]
I'm not able to reproduce with the example application.
The only meaningful difference I've been able to find between my project and the example is that I have different stores for different tile URLs, and I switch back and forth between them. That is when things break.

For now I am just using `.noCache.` and it's fine. I don't need to see the stats very often.

Thanks for the help!
Next I'm gonna test this out on iOS. :crossed_fingers:

**Reply by JaffaKetchup** [reply]
Sure. Excited to see how it goes for you on iOS!

**Reply by eidolonFIRE** [reply]
Seems to be working so far :+1:

**Reply by JaffaKetchup** [reply]
That's great. If you can give a star and a like, and if you can spare any change, that would be greatly appreciated!

**Reply by JaffaKetchup** [reply]
Thanks :)
